### DISCUSSION TITLE
How can create two types of users

### DISCUSSION THREAD

**Discussion by gonta1026** [supabase/supabase#12647] [Questions]
## Problem

For example, I am looking to create two types of users for an e-commerce site application.
One is a user for the admin panel to manage the e-commerce site.
The other is a general user who uses the e-commerce site.

I know I can create one type of user using the authentication feature of supabase, but how can I implement the feature to create the other type of user?
Or is there some way to create two types of users with supabase?

I am thinking of using the following for the technical aspects of the implementation.
- Next.js
- Supabase

**Comment by GaryAustin1** [comment]
Three potential ways, but nothing built in.
1)  Add a role table with user_id primary UUID key and the user's role.  Then manage things using RLS using that table.

2)  Add a role to raw_app_meta_data column in auth.users and use the jwt in the RLS.  https://github.com/supabase-community/supabase-custom-claims

3)  Use Postgres roles and the role column in auth.users.  This used to work and should be working again soon (security patch issues as discussed also in this thread).  https://github.com/supabase/supabase/discussions/11948

**Reply by gonta1026** [reply]
@GaryAustin1 

Thanks for the advice!

> Add a role table with user_id primary UUID key and the user's role. Then manage things using RLS using that table.

I have written what I want to implement now in the following sql, but I am not familiar with sql and it does not work when I do a supabase.auth.signUp.


## What I want to do
- I would like to use a function as a trigger to create users when `new.raw_user_meta_data.role` is admin and when it is not.


```js
    // client code
    const { error } = await supabase.auth.signUp({
      email,
      password,
      options: {
        data: {
          name,
          role: "admin"
        },
      },
    })
```

```sql// supabase function code. This is fired by supabase.auth.signUp.

begin
  if new.raw_user_meta_data.role = 'admin' then
    insert into public.admin_user(id, name) values(
      new.id,
      new.raw_user_meta_data->>'name'
    );
  else
    insert into public.users(id, name) values(
      new.id,
      new.raw_user_meta_data->>'name'
    );  
  end if;
  return new;
end;

```

## image
supabase.auth.signUp to error 

<img width="410" alt="スクリーンショット 2023-02-25 8 43 31" src="https://user-images.githubusercontent.com/45672476/221321773-a49c2266-2871-4678-a339-bdc2c4ecc9b6.png">

**Reply by GaryAustin1** [reply]
You might want to head to Supabase Discord to get more interactive help from users.

Your test is wrong as you are not using json.  `if (new.raw_user_meta_data->>role = 'admin)`  is probably correct.
You can look in the dashboard Postgres logs to get your errors.

Also the user_meta_data is not secure if that call is being made from a browser.  The user can change the role to admin before making the call in browser dev mode.

You need to set admin role in a more secure fashion.

**Reply by gonta1026** [reply]
@GaryAustin1

Thank you very much, I'll check out the discord and see what's going on.

I'm going to check things out on discord, as some of my understanding hasn't caught up with you!

I appreciate all the support you have given me.

https://www.reddit.com/r/Supabase/comments/os2fub/supabase_official_discord_now_open/

**Reply by bakura10** [reply]
Hi @gonta1026 !

i have a similar use case where I have two different typology of users interacting with a system (merchants and customer). If a customer register with their email, they become a customer, but should sign up again if they also want a merchant account.

with Auth0 we have the possibility to create two user databases for that. Is there similar solution for Supabase ?

**Reply by gonta1026** [reply]
@bakura10 
> Is there similar solution for Supabase ?

I don't know 😭 I would like to know if anyone knows.

**Reply by PhilippS93** [reply]
We also need a separation of user pools for our customer/internal admins pools. Admin logins should not be able to log into the customer dashboard with their login and vice versa.

**Reply by jlukas99** [reply]
some updates in this issue?

**Reply by GaryAustin1** [reply]
@jlukas99 They (Supabase) added another way to do role claims in the JWT.  https://supabase.com/docs/guides/database/postgres/custom-claims-and-role-based-access-control-rbac?queryGroups=language&language=plpgsql
Still relies on one sign in for all groups though.

**Reply by liu-ronny** [reply]
@GaryAustin1 is using custom database roles (the third approach) reliable? I asked a related [question](https://github.com/orgs/supabase/discussions/27831) but didn't receive a response.

I currently use roles/permissions tables and custom claims in the JWT, which works well for distinguishing roles within the same user type. However, I'd like a cleaner separation between different user types.

I can think of at least a few benefits of adding a custom database role:
- Clear separation of RLS policies
- Ability to grant different column permissions
- Ability to write functions for different user types and assign execute permissions accordingly

**Reply by GaryAustin1** [reply]
@liu-ronny Well it is reliable as it is all Postgres behind the scenes.  The only thing Supabase involved is the role claim in the jwt and auth.users table.  Supabase cautions about changing the role claim as you can get yourself in trouble.

**Reply by liu-ronny** [reply]
> @liu-ronny Well it is reliable as it is all Postgres behind the scenes.  The only thing Supabase involved is the role claim in the jwt and auth.users table.  Supabase cautions about changing the role claim as you can get yourself in trouble.  

When you say "get yourself in trouble," do you mean:

1. "Do this at your own risk" kind of trouble, or
2. "You're messing with implementation details you're not supposed to touch" kind of trouble?

The client library allows [specifying the role](https://supabase.com/docs/reference/javascript/auth-admin-createuser) when creating or updating a user server-side. I could create an edge function using the JS library, which is documented. But this doesn't seem much different from altering the role column via a Postgres function and letting the Auth server populate new JWTs.

**Reply by GaryAustin1** [reply]
I can't speak for Supabase and it was a general comment in a different discussion from a dev.  I took it as if you don't know what your are doing with Postgres roles you can get your self in trouble.  So 1.  I would not hesitate to do it myself, but I know SQL and the auth schema fairly well.
